_drafts/cortexm-rtos-context-switching.md

```diff
@@ -0,0 +1,996 @@
+---
+title: "A Practical Guide to ARM Cortex-M RTOS Context Switching"
+description: "A deep dive into how the ARM Cortex-M architecture makes multi-tasking with an RTOS
+possible with a step-by-step walk through of the FreeRTOS implementation as an example"
+author: chris
+image: /img/context-switching/pendsv-debug-gdb.png
+---
+
+Many embedded systems reach a level of complexity where having a basic set of scheduling primitives
+and ability to run different tasks can be helpful. The operation of switching from one task to
+another is known as a **context switch**. A _Real Time Operating System_ (**RTOS**) will typically
+provide this functionality. Having a foundational knowledge of how the core of an **RTOS** works
+can be a valuable skill set for an embedded engineer to have.
+
+<!-- excerpt start -->
+
+In this article we will explore how **context switching** works on ARM Cortex-M MCUs. We will
+discuss how the hardware was designed to support this operation, features that impact the context
+switching implementation such as the Floating Point Unit (FPU), and common pitfalls seen when
+porting an RTOS to a platform. We will also walk through a practical example of analyzing the
+**FreeRTOS** context switcher, `xPortPendSVHandler`, utilizing `gdb` to strengthen our understanding.
+
+<!-- excerpt end -->
+
+_Like Interrupt? [Subscribe](http://eepurl.com/gpRedv) to get our latest posts
+straight to your mailbox_
+
+{:.no_toc}
+
+## Table of Contents
+
+<!-- prettier-ignore -->
+* auto-gen TOC:
+{:toc}
+
+{:.no_toc}
+
+## Cortex-M ARM MCU Features
+
+To understand how RTOS context switching works for ARM Cortex-M MCUs, it's critical to have
+foundational knowledge about the primitives the architecture provides to make it possible.
+
+In this section we go through these building blocks by distilling down the information spread
+across the ARM Cortex-M reference manuals and the _ARM
+Architecture Procedure Calling Standard_ (**AAPCS**)[^1] which defines the _Application Binary
+Interface_ (**ABI**) a compiler must abide by for ARM.
+
+> NOTE: If you already have a good understanding of these concepts, feel free to
+> [switch over](#context-switching) this section (pun intended).
+
+{: #cortex-m-privilege-modes}
```

Comment:
Why rename the link?

Reply:
once upon I had an internal link back here, seems like thats no longer the case, removed